primitives/core/src/crypto.rs

```diff
@@ -866,14 +874,33 @@ pub trait Pair: CryptoType + Sized + Clone + Send + Sync + 'static {
 	/// This is generally slower than `generate()`, so prefer that unless you need to persist
 	/// the key from the current session.
 	#[cfg(feature = "std")]
-	fn generate_with_phrase(password: Option<&str>) -> (Self, String, Self::Seed);
+	fn generate_with_phrase(password: Option<&str>) -> (Self, String, Self::Seed) {
+		let mnemonic = Mnemonic::new(MnemonicType::Words12, Language::English);
+		let phrase = mnemonic.phrase();
+		let (pair, seed) = Self::from_phrase(phrase, password)
+			.expect("All phrases generated by Mnemonic are valid; qed");
+		(pair, phrase.to_owned(), seed)
+	}
 
 	/// Returns the KeyPair from the English BIP39 seed `phrase`, or `None` if it's invalid.
 	#[cfg(feature = "std")]
 	fn from_phrase(
 		phrase: &str,
 		password: Option<&str>,
-	) -> Result<(Self, Self::Seed), SecretStringError>;
+	) -> Result<(Self, Self::Seed), SecretStringError> {
+		let big_seed = substrate_bip39::seed_from_entropy(
+			Mnemonic::from_phrase(phrase, Language::English)
+				.map_err(|_| SecretStringError::InvalidPhrase)?
+				.entropy(),
+			password.unwrap_or(""),
+		)
+		.map_err(|_| SecretStringError::InvalidSeed)?;
+		let mut seed = Self::Seed::default();
+		let seed_slice = seed.as_mut();
+		let minlen = seed_slice.len().min(big_seed.len());
+		seed_slice[0..minlen].copy_from_slice(&big_seed[0..minlen]);
```

Comment:
Yeah probably would be better to fail than filling the seed with zeroes.

What about embedding the unique function from the `substrate-bip39` directly in substrate?
As I said that library provides 1 function (the other is not used anymore after this pr).

The function is this:
```
pub fn seed_from_entropy(entropy: &[u8], password: &str) -> Result<[u8; 64], Error> {
    if entropy.len() < 16 || entropy.len() > 32 || entropy.len() % 4 != 0 {
        return Err(Error::InvalidEntropy);
    }
    let mut salt = String::with_capacity(8 + password.len());
    salt.push_str("mnemonic");
    salt.push_str(password);
    let mut seed = [0u8; 64];
    pbkdf2::<Hmac<Sha512>>(entropy, salt.as_bytes(), 2048, &mut seed);
    salt.zeroize();
    Ok(seed)
}
```

Sounds a bit bombastic to define a library only to provide a function...

Furthermore, if we can then easily modify it to fill a `mut slice` (passed as parameter) to fill it with an arbitrary number of bytes... Thus there would be no need to check if we can fill the seed buffer since we always can

Reply:
I have no real opinion here. We can copy the function, I don't really care. I don't know the history behind the substrate-bip39 crate.